core/src/main/java/org/infinispan/partitionhandling/impl/PreferAvailabilityStrategy.java

```diff
@@ -93,166 +68,283 @@ public void onClusterViewChange(AvailabilityStrategyContext context, List<Addres
+      for (int i = 0; i < partitions.size();) {
+         Partition p = partitions.get(i);
 
-      // ISPN-8903 attempt CR in the event that a paused/busy process resumes a larger CH then the previous coordinator
-      return maxTopologyId > maxTopology.getTopologyId();
+         boolean fold = false;
+         for (int j = 0; j < i; j++) {
+            Partition referencePartition = partitions.get(j);
+
+            // If read owners don't overlap then we clearly have 2 independent partitions
+            if (!InfinispanCollections.containsAny(p.readCH.getMembers(), referencePartition.readCH.getMembers())) {
+               continue;
+            }
+
+            // Normally the difference between the topology ids seen by different members is 0 or 1.
+            // But only rebalance start and phase changes are kept in step,
+            // topology updates are fire-and-forget, so the difference can be higher
+            // even when communication between nodes is not interrupted.
+            int referenceTopologyId = referencePartition.topology.getTopologyId();
+            int topologyId = p.topology.getTopologyId();
+            if (topologyId == referenceTopologyId) {
+               if (p.topology.equals(referencePartition.topology)) {
+                  // The topology is exactly the same, ignore it
+                  log.tracef("Cache %s ignoring topology from %s, same as topology from %s: %s",
+                             cacheName, p.senders, referencePartition.senders, p.topology);
+                  fold = true;
+               } else {
+                  if (trace)
+                     log.tracef(
+                        "Cache %s partition of %s overlaps with partition of %s, with the same topology id",
+                        cacheName, p.senders, referencePartition.senders);
+               }
+            } else {
+               // topologyId < referenceTopologyId
+               // Small topology differences are expected even between properly communicating nodes,
+               // but we should consider it a different partition if p's sender might have some entries
+               // that the other members lost.
+               // Having a majority doesn't matter, because the lost data check ensures p's sender
+               // couldn't make any updates without talking to the reference partition's members
+               if (!lostDataCheck.test(p.readCH, referencePartition.actualMembers)) {
+                  log.tracef("Cache %s ignoring compatible old topology from %s: %s", cacheName, p.senders, p.topology);
+                  fold = true;
+               } else {
+                  if (trace)
+                     log.tracef("Cache %s partition of %s overlaps with partition of %s but possibly holds extra entries",
+                                cacheName, p.senders, referencePartition.senders);
+               }
+            }
+            if (fold) {
+               referencePartition.senders.addAll(p.senders);
```

Comment:
I kinda wonder if this would be simpler if the outer i loop was doing `for (int i = 0; i < partitions.size(); ++i)` and we removed the `referencePartition` instead?

Reply:
Done